tests/mock_network/utils.rs

```diff
@@ -743,38 +802,137 @@ pub fn verify_invariants_for_nodes(env: &Environment, nodes: &[TestNode]) {
     }
 }
 
+// Send an `UserMessage` with `content` from `src` to `dst`.
+pub fn send_user_message(
+    nodes: &mut [TestNode],
+    src: Prefix<XorName>,
+    dst: Prefix<XorName>,
+    content: Vec<u8>,
+) {
+    let src_location = SrcLocation::Section(src);
+    let dst_location = DstLocation::Section(dst.name());
+
+    trace!(
+        "send_user_message: {:?} -> {:?}: {:?}",
+        src_location,
+        dst_location,
+        hex_fmt::HexFmt(&content),
+    );
+
+    for node in elders_with_prefix_mut(nodes, &src) {
+        node.inner
+            .send_message(src_location, dst_location, content.clone())
+            .unwrap()
+    }
+}
+
+// Poll until all sections have up-to-date knowledge of their neighbour sections.
+// We consider section A's knowledge of its neighbour section B as up-to-date if every elder from A
+// knows at least `threshold` online nodes from B.
+pub fn update_neighbours_and_poll(env: &Environment, nodes: &mut [TestNode], threshold: usize) {
+    info!("update_neighbours_and_poll start");
+
+    let mut rng = env.new_rng();
+
+    // Max number of times we will try to send messages and wait until the knowledge gets updated.
+    let max_sends = 3;
+    // Number of iterations until we send new messages.
+    let mut send_countdown = 0;
+
+    for _ in 0..POLL_UNTIL_MAX_ITERATIONS {
+        let outdated: BTreeSet<_> = neighbours_with_outdated_knowledge(nodes, threshold).collect();
+        if outdated.is_empty() {
+            info!("update_neighbours_and_poll done");
+            return;
+        }
+
+        if send_countdown == 0 {
+            send_countdown = POLL_ALL_MAX_ITERATIONS / max_sends;
+
+            for (a, b) in outdated {
+                let content = gen_vec(&mut rng, 32);
+                send_user_message(nodes, a, b, content);
+            }
+        } else {
+            send_countdown -= 1;
+        }
+
+        poll_all(env, nodes);
+        advance_time(POLL_UNTIL_TIME_STEP);
+    }
+
+    panic!(
+        "update_neighbours_and_poll failed after {} iterations",
+        POLL_UNTIL_MAX_ITERATIONS
+    );
+}
+
+// Returns iterator over pairs of neighbour sections where at least one has outdated knowledge of
+// the other. See `section_knowledge_is_up_to_date` for the meaning of `threshold`.
+fn neighbours_with_outdated_knowledge<'a>(
+    nodes: &'a [TestNode],
+    threshold: usize,
+) -> impl Iterator<Item = (Prefix<XorName>, Prefix<XorName>)> + 'a {
+    let prefixes: Vec<_> = current_sections(nodes).collect();
+    prefixes
+        .into_iter()
+        .tuple_combinations()
+        .filter(|(a, b)| a.is_neighbour(b))
+        .filter(move |(a, b)| {
+            !section_knowledge_is_up_to_date(nodes, a, b, threshold)
+                || !section_knowledge_is_up_to_date(nodes, b, a, threshold)
+        })
+        .map(|(a, b)| if a < b { (a, b) } else { (b, a) })
```

Comment:
I think when there is an out-of-date happens, there will be `(a, b)` and `(b, a)` both be reported. 
However, with this ordering, it means within the iterator, there will be two same entries ?

Reply:
No, because `tuple_combinations` ignores order: https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.tuple_combinations

The ordering is actually unnecessary now. I put it there to make sure the function doesn't return `(a, b)` once and `(b, a)` the second time, but then I changed the code that required it. Will remove it to simplify the code and avoid confusion.

Reply:
ah, good to know that :)